inject-java/src/main/java/io/micronaut/annotation/processing/TypeElementVisitorProcessor.java

```diff
@@ -228,18 +229,21 @@ public boolean process(Set<? extends TypeElement> annotations, RoundEnvironment
 
             TypeElement groovyObjectTypeElement = elementUtils.getTypeElement("groovy.lang.GroovyObject");
             TypeMirror groovyObjectType = groovyObjectTypeElement != null ? groovyObjectTypeElement.asType() : null;
+            Predicate<TypeElement> notGroovyObject = typeElement -> groovyObjectType == null || !typeUtils.isAssignable(typeElement.asType(), groovyObjectType);
 
             Set<TypeElement> elements = new LinkedHashSet<>();
 
             for (TypeElement annotation : annotations) {
-                final Set<? extends Element> annotatedElements = roundEnv.getElementsAnnotatedWith(annotation);
-                includeElements(elements, annotatedElements, groovyObjectType);
+                modelUtils.resolveTypeElements(
+                    roundEnv.getElementsAnnotatedWith(annotation)
+                ).filter(notGroovyObject).forEach(elements::add);
             }
 
             // This call to getRootElements() should be removed in Micronaut 4. It should not be possible
             // to process elements without at least one annotation present and this call breaks that assumption.
-            final Set<? extends Element> rootElements = roundEnv.getRootElements();
-            includeElements(elements, rootElements, groovyObjectType);
+            modelUtils.resolveTypeElements(
```

Comment:
> // This call to getRootElements() should be removed in Micronaut 4. It should not be possible
> // to process elements without at least one annotation present and this call breaks that assumption.
> 

I left it as it is because some tests will fail if this is removed.